### DISCUSSION TITLE
How to ensure that each particle's ID is unique during parallel execution?

### DISCUSSION THREAD

**Discussion by 1120805582** [BLAST-WarpX/warpx#4860] [Data Analysis] [answered]
Hello, I would like to inquire about tracking particles. How to ensure that the range of particle IDs assigned to each process is mutually exclusive during parallel execution? It appears that at the initial moment, a very large number of particles（about 64*64） shared the same ID value.
![download](https://github.com/ECP-WarpX/WarpX/assets/112069701/c39e581d-fe92-4dcd-9fb6-443e2f25ee7c)

**Comment by ax3l** [comment, marked as answer]
Hi @1120805582, can you clarify how you are plotting exactly? (Python snippet?)
Do you plot this while running via Python WarpX (PICMI) or from output data (plotfiles? openPMD?)?

By default, our 64bit particle indices are unique and not reassigned over simulation lifetime, including parallel execution.

We do this by varying the high bits of each MPI rank for the ids we generate, which we call the `cpu` part. We then combine this with lower bits, which we call the _MPI-local_ `id` part in AMReX and which is not unique in itself. Combined, this makes the parallelly unique `idcpu`, which is 64bit.

**Reply by 1120805582** [reply]
Thanks for your assistance with the tracking of particle trajectories using particle ID and CPU identification. This method has successfully allowed me to follow specific particles within my simulations.

However, I have encountered a challenge in scenarios where particles are continuously generated. It appears that even the combination of CPU and particle ID does not maintain consistent identification of the same particle. Instead, the trajectories associated with a specific CPU and ID combination seem to be discontinuous and unordered, with abrupt changes in energy levels.

Could you provide any advice on how to effectively track individual particles during continuous generation periods? I am currently using version 23.03 of WarpX.

Thank you very much for your attention to this matter.
